Every task is a loop in the thread pool

dsharlet opened the pull request:
This makes every queued task in the thread pool a `parallel_for<>` object instead of a simple task. (Tasks are now loops with `n = 1`.)

This eliminates the need for the `cancel` mechanism, which was clunky and high overhead. Now, we can simply see if any work remains *before* starting to work on a task instead of after, which eliminates a lot of unneeded worker tasks, especially in the case of nested parallel loops.

vksnk reviewed (approved):
I think this needs a general comment explaining what does it mean for  the thread pool to operate on loops, how are they different from tasks, maybe some simple textual example of the execution workflow.

dsharlet commented:
I added a small bit to one of the comments, but overall I don't think this actually changes the interface much. The main difference is how tasks get dequeued from the task queue internally. The one difference is where I added some more comments.
